src/GraphQL/Utilities/MetadataProvider.cs

```diff
@@ -11,24 +10,67 @@ namespace GraphQL.Utilities
     /// </summary>
     public class MetadataProvider : IProvideMetadata
     {
+        private AppliedDirectives _appliedDirectives;
+        private Dictionary<string, object> _metadata;
+
         /// <inheritdoc />
-        public IDictionary<string, object> Metadata { get; set; } = new ConcurrentDictionary<string, object>();
+        public AppliedDirectives AppliedDirectives
+        {
+            get
+            {
+                if (_appliedDirectives == null)
+                {
+                    lock (this)
+                    {
+                        if (_appliedDirectives == null)
+                        {
+                            _appliedDirectives = new AppliedDirectives();
+                        }
+                    }
+                }
+
+                return _appliedDirectives;
+            }
+        }
+
+        /// <inheritdoc />
+        public Dictionary<string, object> Metadata
+        {
+            get
+            {
+                if (_metadata == null)
+                {
+                    lock (this)
+                    {
+                        if (_metadata == null)
+                        {
+                            _metadata = new Dictionary<string, object>();
+                        }
+                    }
+                }
+
+                return _metadata;
+            }
+        }
 
         /// <inheritdoc />
         public TType GetMetadata<TType>(string key, TType defaultValue = default)
         {
-            var local = Metadata;
+            var local = _metadata;
             return local != null && local.TryGetValue(key, out object item) ? (TType)item : defaultValue;
         }
 
         /// <inheritdoc />
         public TType GetMetadata<TType>(string key, Func<TType> defaultValueFactory)
         {
-            var local = Metadata;
+            var local = _metadata;
             return local != null && local.TryGetValue(key, out object item) ? (TType)item : defaultValueFactory();
         }
```

Comment:
It would be nice to change these methods to a single `bool TryGetMetadata(string key, out TType item)` method, perhaps keeping these temporarily for v4.x and marking them as deprecated.

Reply:
`GetMetadata` actually already works as `TryGetMetadata`

Reply:
Only if you can provide a default value that isn’t valid for the dictionary item. That’s why a dictionary provides TryGetValue, so it can conclusively be determined that a value has never been set. And we don’t need an overload with a factory.

Reply:
Well, maybe, but not now.